### ISSUE TITLE
[osm] cookie control

### ISSUE THREAD

**Issue by brianteeman** [joomla/joomla-websites#1425]
Am I correct in assuming that the cookie control script on the osm web site is the one that it is intended to role out on the joomla.org websites ?

If so then there are quite a lot of accessibility issues with it and I would like to know the correct place to report them.

**Comment by jeckodevelopment** [comment]
Hi Brian,
you're correct, that was the tool that is going to be deployed.
I'd be grateful if you can summarize the accessibility issues here in this issue.
Thanks

**Comment by brianteeman** [comment]
It would be more productive if you could let me know where the code is and then I can comment more constructively with actual fixes. The original script is accessible so either it is in some of the options you have chosen or some customisations that have broken things

**Comment by conconnl** [comment]
The version which is being tested runs on https://test.volunteers.joomla.org/
A PR is already created on the template Repository and Michael, David and others have done review a requested multiple changes, which need to be implemented and test on the mentioned site again. Including some display things. The one on OSM site is an old version.

**Comment by brianteeman** [comment]
@jeckodevelopment As requested

1. 71 tab stops for a keyboard user to reach the accept button
2. Screen readers have a bit more luck as they are told about an access key. But when they use it they hear
> This website uses cookies to remember users and understand ways to enhance their experience. CLOSE
3. There is an unlabelled checkbox 
![image](https://user-images.githubusercontent.com/1296369/83519562-1f195f80-a4d4-11ea-9f6a-f02812c4761c.png)
4. Configure Advendors
![image](https://user-images.githubusercontent.com/1296369/83519830-8df6b880-a4d4-11ea-861b-267fed4dda30.png)
The information on the following page is not about configuring advendors. It is the page for configuring the type of cookies you will accept
5. If you have already accepted the cookies then the access key is not announced on page load anymore and you will have to navigate through more than 70 links to find the control to change your cookie settings.
6. The version of cookie control being used was end of life in March

**Comment by jeckodevelopment** [comment]
Thanks @brianteeman 
@Alcaeusd can you look at that?

**Comment by Alcaeusd** [comment]
> Thanks @brianteeman
> @Alcaeusd can you look at that?

@jeckodevelopment  The changes needed for the cookie script will be applied soon and then I will take a look into this. TBH I will need some help on setting up any necessary attributes or fixing any unwanted behavior as @brianteeman mentioned. The version will updated too.

**Comment by conconnl** [comment]
@Alcaeusd @dkalnenaite @jeckodevelopment 
When is the @joomla/privacy-compliance-team going to finish the missing Cookie Control implementation ?

**Comment by PhilETaylor** [comment]
The OSM website is now displaying this in the console.log 


> We would like to inform Cookie Control users that IAB has announced the end of support for TCFv1.1 at the end of March 2020. Full support for IAB TCFv2.0 and CCPA will be included in Cookie Control v9.0, which all users will be eligible to download for free once it is released. Customers who use the IAB functionality are strongly advised to plan for this upgrade as soon as possible. For more information, please visit https://www.civicuk.com

**Comment by conconnl** [comment]
> The OSM website is now displaying this in the console.log
> 
> > We would like to inform Cookie Control users that IAB has announced the end of support for TCFv1.1 at the end of March 2020. Full support for IAB TCFv2.0 and CCPA will be included in Cookie Control v9.0, which all users will be eligible to download for free once it is released. Customers who use the IAB functionality are strongly advised to plan for this upgrade as soon as possible. For more information, please visit https://www.civicuk.com

Thanks Phil.
I have implemented the update in the OSM website.
For what I can see, all is working and the console doesn't show errors.
Could you do a check as well?

**Comment by PhilETaylor** [comment]
> Could you do a check as well?

The only message I get now, is the message all Joomla sites exhibit

`JQMIGRATE: Migrate is installed, version 1.4.1`

**Comment by dkalnenaite** [comment]
Hi all,

Just to clarify, the CCPA does not apply to us.

Best,
Donata Kalnenaite

**Comment by brianteeman** [comment]
As its been over a year since the closed source proprietary cookie script was first implemented on one site and it still has not been implemented across any of the other sites and it fails on accessibility is there any interest in implementing a 100% free MIT licenced cookie control script that offers similar (perhaps better) functionality. It is actively developed and used by many government web sites in the EU and I can deliver it as a joomla plugin within 1 week

**Comment by dkalnenaite** [comment]
@brianteeman what is the name of the tool? Can you send us a link to the Privacy Policy of the tool if there is one?

**Comment by brianteeman** [comment]
https://github.com/KIProtect/klaro

**Comment by dkalnenaite** [comment]
@brianteeman is there a Privacy Policy for this tool? 

Below are the compliance requirements that the cookie consent tool needs to meet. Can you please confirm whether this tool meets these requirements? 
1. Have all non-essential cookies disabled by default; 
2. Allow individuals to choose which cookies they'd like to accept - both by type (e.g. marketing cookies) and by specific cookie (e.g. Google Analytics); 
3. Give users a real choice and not nudge them towards a particular choice. For example, the "decline" button should be as prominent as the "accept" button - i.e. the "decline" button should not be hidden; 
4. Track what choices the user has made; 
5. Allow the user to retract their consent at any time.

**Comment by brianteeman** [comment]
Can you explain what you mean by 4

**Comment by sebastienlapoux** [comment]
IMO 4) Track what choices the user has made -> to bring the proof of the validity of user consents. (a solution is to record consents of course we none personal data...  OneTrust is proposing this)

**Comment by brianteeman** [comment]
surely we cant store that

**Comment by dkalnenaite** [comment]
When a user consents to cookies, we need to be able to show that they consented to being tracked with those cookies. Otherwise, we have no proof that we complied with the law.

**Comment by brianteeman** [comment]
I had never heard of people maintaining an audit trail of cookie consents.  Certainly the current solution doesnt offer that and neither does the one I am suggesting.

I dont even see how it is even possible for non-authenticated users

**Comment by zero-24** [comment]
> When a user consents to cookies, we need to be able to show that they consented to being tracked with those cookies. Otherwise, we have no proof that we complied with the law.

Isnt the proofe that we dont include any scripts etc once approved by the user. I'm not sure whether having a list of IPs would be allowed anyway.

**Comment by conconnl** [comment]
I just checked and it seems that the Cookie Control from CIVIC has this possibility, but has never been enabled.
If this would be a requirement, I don't have a clue why the implementation wasn't done.

The recording is done in the system of the Third Party, which brings other legal consequences.
At the same time, what they record doesn't give you any real audit trail.
They will give you a date, an anonymized IP address, the website address from where the request was received, the user agent string and the UUID of the browser.

Recording doesn't bring any real usable information, so there is no reason to use it.
IP Addresses can change, Agent String is the same for a specific browser and version and UUID can also change if you reinstall or empty the in-browser cache

What @zero-24 wrote:
> Isnt the proofe that we dont include any scripts etc once approved by the user. I'm not sure whether having a list of IPs would be allowed anyway.

If we can prove our system works, there is no reason to record what a user clicked.
Recording that they consent, while they can unconsent a minute later, makes the record false.

**Comment by brianteeman** [comment]
> Recording doesn't bring any real usable information, so there is no reason to use it.

exactly

**Comment by brianteeman** [comment]
Anyway the offer is on the table to actually comply with the law and have a cookie control script ready to use on our web sites within 1 week - we're already several years non-compliant for no good reason

**Comment by conconnl** [comment]
I fully agree....
@brianteeman correct me if I miss understand their system / website

Looking at this system it doesn't require a relation with the system of the supplier at all.
Which is required with the one from CIVIC.
We don't need a privacy policy, data processor agreement and such things.

**Comment by brianteeman** [comment]
> Looking at this solution it doesn't require a relation with the system of the supplier at all.

Correct - it is a completely staandalone and truly opensource solution

**Comment by sebastienlapoux** [comment]
The data controllers must be able to demonstrate that the user has given his consent.

Sorry not the time to bring you all justifications of my recommendations below:
* record consents at least with date, object of the consent, user's consent (y or n) -> about to 
provide individual evidence of the collection of user consent
* placing the code used by the organization collecting the consent, for the different versions of its site or mobile application in an escrow managed by a third party

Don't record the IP anonymized or not.

**Comment by brianteeman** [comment]
> The data controllers must be able to demonstrate that the user has given his consent.

How is that possible on a non-authenticated system.

**Comment by zero-24** [comment]
And without using the IP. ;-)

**Comment by dkalnenaite** [comment]
Hi all, 

Below is some information on how consent needs to be recorded under GDPR. I understand that we need a cookie consent solution to comply with GDPR. However, if that solution is unable to correctly record consent, then it is not going to help us comply with GDPR. 

How should we record consent?
Article 7(1) says:

“Where the processing is based on consent, the controller shall be able to demonstrate that the data subject has consented to process of his or her personal data.”

This means you must have an effective audit trail of how and when consent was given, so you can provide evidence if challenged. You should keep this evidence for as long as you are still processing based on the consent so that you can demonstrate your compliance in line with accountability obligations.

Good records will also help you to monitor and refresh consent as appropriate. You must keep good records that demonstrate the following:

Who consented: the name of the individual, or other identifiers (eg, online user name, session ID).
When they consented: a copy of a dated document or online records that include a timestamp; or, for oral consent, a note of the time and date which was made at the time of the conversation.
What they were told at the time: a master copy of the document or data capture form containing the consent statement in use at that time, along with any separate privacy policy or other privacy information, including version numbers and dates matching the date consent was given. If consent was given orally, your records should include a copy of the script used at that time.
How they consented: for written consent, a copy of the relevant document or data capture form. If consent was given online, your records should include the data submitted as well as a timestamp to link it to the relevant version of the data capture form. If consent was given orally, you should keep a note of this made at the time of the conversation - it doesn’t need to be a full record of the conversation.
Whether they have withdrawn consent: and if so, when.

Examples of proper recording of consent are as follows: 

- You keep a copy of the customer’s signed and dated form that shows they ticked to provide their consent to the specific processing.
- You keep records that include an ID and the data submitted online together with a timestamp. You also keep a copy of the version of the data-capture form and any other relevant documents in use at that date.
- You keep records that include the time and date of the conversation, the name and date/version of the script used.

Consent should be specific and granular, so your records also need to be specific and granular to demonstrate exactly what the consent covers.

For online consent, you may be able to use an appropriate cryptographic hash function to support data integrity.

**Comment by brianteeman** [comment]
That is the regulation for GDPR and the storing of user data.

This issue is about cookie.

It is completely impossible to achieve what you describe for a cookie

**Comment by conconnl** [comment]
I fully agree.
It's technically not possible to store who gave consent for which cookie.
Specifically the who is impossible.

The cookies consent happen based on your visit, depend on the browser you use for that moment, if you allow cookies to stay in your browser storage after closing. If you use containers in your browser split many technical things if needed.
Happen before your authenticate, fill forms etc.
The information you want to store for the consent is invalid the moment you store it, as it can change at the same moment.

An IP address won't help, people in the same location use the same IP address, thus it can be multiple persons.
You have to anonymize the IPs so it makes the info more absolute and not related to a person.
Other information you don't have available when clicking "Accept" for cookies, so there is nothing to relate to a person and thus prove who it is and if they consented.
The only way you can proof that people consent to it, is the technical implementation of a solution.
How it works, if it really doesn't place the needed cookies and so on.

**Comment by brianteeman** [comment]
https://www.osano.com/articles/cookies-proof-of-consent-records-of-consent#:~:text=The%20cookie%20law%20requires%20users,device%20and%2For%20tracking%20them.&text=The%20Cookie%20Law%20does%20not,that%20consent%20has%20been%20withdrawn

**Comment by brianteeman** [comment]
Closed and withdrawing the offer. 
There clearly is a lack of interest from leadership in finally making the joomla.org properties compliant with the law while we promote a proprietary service which goes against everything joomla should believe in.
Its a shame that people would prefer to write long documents instead of actually implementing anything.
There is no logical reason why something that shouldnt take more than a week instead has taken several years already and without any progress in being compliant.

**Comment by dkalnenaite** [comment]
@brianteeman sorry for my late response to this - I've spent the last week with a bad toothache. If there is no way to store the cookie consent, then that's fine. We can implement the best solution that is available and go from there with improvements in the future if those are necessary. I'm not sure if the "leadership" comment is referring to me or to our DPO but I do not believe that I have the authority to approve or disapprove any compliance solutions. I was just providing my opinion.
